File: Data Structures & Algorithms/three-integer-sum/submission-0.js
class Solution {
    threeSum(nums) {
        nums.sort((a,b)=>a-b);
        let result = [];

        for (let i = 0; i < nums.length; i++) {

            if (i > 0 && nums[i] === nums[i-1]) continue;

            let p1 = i + 1;
            let p2 = nums.length - 1;
            let target = -nums[i];

            while (p1 < p2) {
                let sum = nums[p1] + nums[p2];

                if (sum === target) {
                    result.push([nums[i], nums[p1], nums[p2]]);

                    p1++;
                    p2--;

                    while (p1 < p2 && nums[p1] === nums[p1-1]) p1++;
                    while (p1 < p2 && nums[p2] === nums[p2+1]) p2--;
                }
                else if (sum < target) {
                    p1++;
                }
                else {
                    p2--;
                }
            }
        }
        return result;
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:

    def _find_combi(self, nums, target, combi) -> List[List[int]]:
        if target == 0:
            return [combi]
        if not nums or target < 0:
            return []

        result = []
        n = nums[0]

        for k in range((target // n + 1)):
            result.extend(self._find_combi(nums[1:], target - k * n, combi + [n] * k))

        return result



    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        return self._find_combi(nums, target, [])

            
            
            



        



        